rasa/telemetry.py

```diff
@@ -0,0 +1,553 @@
+        "properties": properties,
+        "context": context,
+    }
+
+
+def in_continuous_integration() -> bool:
+    """Returns `True` if currently running inside a continuous integration context."""
+    return any(env in os.environ for env in CI_ENVIRONMENT_TELL)
+
+
+def _is_telemetry_debug_enabled() -> bool:
+    """Check if telemetry debug mode is enabled."""
+    return (
+        os.environ.get(TELEMETRY_DEBUG_ENVIRONMENT_VARIABLE, "false").lower() == "true"
+    )
+
+
+def print_telemetry_event(payload: Dict[Text, Any]) -> None:
+    """Print a telemetry events payload to the commandline.
+
+    Args:
+        payload: payload of the event
+    """
+    print("Telemetry Event:")
+    print(json.dumps(payload, indent=2))
+
+
+async def _send_event(
+    distinct_id: Text,
+    event_name: Text,
+    properties: Dict[Text, Any],
+    context: Dict[Text, Any],
+) -> None:
+    """Report the contents of an event to the /track Segment endpoint.
+    Documentation: https://segment.com/docs/sources/server/http/
+
+    Do not call this function from outside telemetry.py! This function does not
+    check if telemetry is enabled or not.
+
+    Args:
+        distinct_id: Unique telemetry ID.
+        event_name: Name of the event.
+        properties: Values to report along the event.
+        context: Context information about the event.
+    """
+
+    payload = segment_request_payload(distinct_id, event_name, properties, context)
+
+    if _is_telemetry_debug_enabled():
+        print_telemetry_event(payload)
+        return
+
+    write_key = telemetry_write_key()
+    if not write_key:
+        # If TELEMETRY_WRITE_KEY is empty or `None`, telemetry has not been
+        # enabled for this build (e.g. because it is running from source)
+        logger.debug("Skipping request to external service: telemetry key not set.")
+        return
+
+    headers = segment_request_header(write_key)
+
+    async with aiohttp.ClientSession() as session:
+        async with session.post(
+            SEGMENT_ENDPOINT, headers=headers, json=payload,
+        ) as resp:
+            # handle different failure cases
+            if resp.status != 200:
+                response_text = await resp.text()
+                logger.debug(
+                    f"Segment telemetry request returned a {resp.status} response."
+                    f"Body: {response_text}"
+                )
+            else:
+                data = await resp.json()
+                if not data.get("success"):
+                    logger.debug(
+                        f"Segment telemetry request returned a failure."
+                        f"Response: {data}"
+                    )
+
+
+def _project_hash() -> Text:
+    """Create a hash for the project in the current working directory.
+
+    Returns:
+        project hash
+    """
+    try:
+        remote = check_output(  # skipcq:BAN-B607,BAN-B603
+            ["git", "remote", "get-url", "origin"], stderr=STDOUT
```

Comment:
Why do we want to capture the error output? This should fail silently, shouldn't it?

Reply:
in examples usually, it gets pipped there if you don't want errors to end up in your error stream. I changed it to pipe it to dev null.